packages/phalanx/example/FiniteElementAssembly_MixedFieldTypes/evaluators/GatherSolution_Def.hpp

```diff
@@ -0,0 +1,204 @@
+               const int& in_field_index,
+               const Kokkos::View<double*,PHX::Device>& in_x) :
+  num_equations(in_num_equations),
+  field_index(in_field_index),
+  x(in_x)
+{
+  PHX::Tag<ScalarT> field_tag(field_name,layout);
+  this->addEvaluatedField(field_tag,field);
+  this->setName("Gather Solution: "+field_tag.name());
+}
+
+// **********************************************************************
+template<typename Traits>
+void GatherSolution<PHX::MyTraits::Residual, Traits>::
+evaluateFields(typename Traits::EvalData workset)
+{
+  gids = workset.gids_;
+  cell_global_offset_index = workset.first_cell_global_index_;
+  Kokkos::parallel_for(Kokkos::TeamPolicy<PHX::exec_space>(workset.num_cells_,Kokkos::AUTO()),*this);
+}
+
+// **********************************************************************
+template<typename Traits>
+void GatherSolution<PHX::MyTraits::Residual,Traits>::
+operator()(const Kokkos::TeamPolicy<PHX::exec_space>::member_type& team) const
+{
+  const int cell = team.league_rank();
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team,0,field.extent(1)), [=] (const int& node) {
+      field(cell,node) = x( gids(cell_global_offset_index+cell,node) * num_equations + field_index);
+  });
+}
+
+// **********************************************************************
+// Specialization: Jacobian
+// **********************************************************************
+
+template<typename Traits>
+GatherSolution<PHX::MyTraits::Jacobian, Traits>::
+GatherSolution(const std::string& field_name,
+               const Teuchos::RCP<const PHX::DataLayout>& layout,
+               const int& in_num_equations,
+               const int& in_field_index,
+               const Kokkos::View<double*,PHX::Device>& in_x) :
+  num_equations(in_num_equations),
+  field_index(in_field_index),
+  x(in_x)
+{
+  PHX::Tag<ScalarT> field_tag(field_name,layout);
+  this->addEvaluatedField(field_tag,field);
+  this->setName("Gather Solution: "+field_tag.name());
+}
+
+// **********************************************************************
+template<typename Traits>
+void GatherSolution<PHX::MyTraits::Jacobian, Traits>::
+evaluateFields(typename Traits::EvalData workset)
+{
+  gids = workset.gids_;
+  cell_global_offset_index = workset.first_cell_global_index_;
+  Kokkos::parallel_for(Kokkos::TeamPolicy<PHX::exec_space>(workset.num_cells_,Kokkos::AUTO()),*this);
+}
+
+// **********************************************************************
+template<typename Traits>
+void GatherSolution<PHX::MyTraits::Jacobian,Traits>::
+operator()(const Kokkos::TeamPolicy<PHX::exec_space>::member_type& team) const
+{
+  const int cell = team.league_rank();
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team,0,field.extent(1)), [=] (const int& node) {
+      field(cell,node).val() = x(gids(cell_global_offset_index+cell,node) * num_equations + field_index);
+      field(cell,node).fastAccessDx(num_equations * node + field_index) = 1.0;
+  });
+}
+
+/*
```

Comment:
I was keeping the commented out code around. It is for a third evaluation type for Jacobian-vector products. I plan to revive that at some point. We used that code in a paper for an AD conference many years ago.